Allow compressed metadata in flight

jku opened the pull request:
from commit description:
> This commit tries to deal with two interests:
>     * metadata is highly repetitive and compressible: allowing compression
>       would be good
>     * there may be broken web servers (see
>       https://github.com/pypa/pip/blob/404838abcca467648180b358598c597b74d568c9/src/pip/_internal/download.py#L842)
>       that have problems with compression on already compressed target files
>     
>  We can make things better for that first interest while we have no real
>     data for the second interest -- our current workarounds to avoid
>     compression are based on hearsay, not testing.
>     
>  Now that individual fetchers are possible I suggest we simplify
>     ngclient and allow compression. As an example the pip Fetcher
>     could still use the pip response chunking code with all their
>     workarounds -- pip certainly has better capability to maintain
>     a mountain of workarounds and also has endless amounts of real-world
>     testing compared to python-tuf.
>     
>  Details:
>     * Stop modifying Accept-Encoding (Requests default includes gzip)
>     * Don't use response.raw in RequestsFetcher as there is no need anymore
>     * Fix issue in test_session_get_timeout(): it's not mocking the error that requests really raises in this case
> 


Fixes #1251

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
tests: Remove unused options from simple_server

We never call simple_server with arguments so this is dead code.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
ngclient: allow compression in HTTP responses

This commit tries to deal with two interests:
* metadata is highly repetitive and compressible: allowing compression
  would be good
* there may be broken web servers (see
  https://github.com/pypa/pip/blob/404838abcca467648180b358598c597b74d568c9/src/pip/_internal/download.py#L842)
  that have problems with compression on already compressed target files

We can make things better for that first interest while we have no real
data for the second interest -- our current workarounds to avoid
compression are based on hearsay, not testing.

Now that individual fetchers are possible I suggest we simplify
ngclient and allow compression. As an example the pip Fetcher
could still use the pip response chunking code with all their
workarounds -- pip certainly has better capability to maintain
a mountain of workarounds and also has endless amounts of real-world
testing compared to python-tuf.

Details:
* Stop modifying Accept-Encoding (Requests default includes gzip)
* Don't use response.raw in RequestsFetcher as there is no need:
  This was a workaround for false "Content-encoding: gzip" inserted by
  a broken server -- and the workaround was only possible because we
  knew we never asked for compression
* Fix issue in test_session_get_timeout(): it's not mocking the error
  that requests really raises in this case

Fixes #1251

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

kairoaraujo commented:
LGTM
I'm not sure about the third commit as well.
As we talked about before, ``requests`` support it, so we are testing something outside the TUF domain.

jku commented:
Marking non-draft with only these changes:
* rebased on develop
* removed the last commit (_tests: Add gzip support to simple_server_): we don't want to test every Requests feature that is essentially opaque to us. The commit is still conserved in another branch: https://github.com/jku/python-tuf/tree/gzip-simple-server

lukpueh commented on tests/simple_server.py:
```diff
@@ -3,64 +3,16 @@
 socketserver.TCPServer.allow_reuse_address = True
 
-httpd = socketserver.TCPServer(("localhost", 0), handler)
+httpd = socketserver.TCPServer(("localhost", 0), SimpleHTTPRequestHandler)
```
👍  Thanks for cleaning up!

As the comment above says, I introduced this workaround for testing on Windows in Python2 (https://github.com/theupdateframework/python-tuf/commit/7dbb30ae1006d0b5f506bba19d4b53398c44f139). So this has become irrelevant quite a while ago.

lukpueh commented on tuf/ngclient/_internal/requests_fetcher.py:
```diff
@@ -99,26 +99,12 @@ def _chunks(self, response: "requests.Response") -> Iterator[bytes]:
-                if not data:
-                    break
-
+            for data in response.iter_content(self.chunk_size):
```
Could you remind me why we used to read from `response.raw`,  why it is not necessary anymore, and how it is related to compression?

lukpueh replied:
Do we give up any needed control about the incoming data stream?

jku replied:
oh, great question, should have talked about that.

I believe this was added to pip for this case:
* broken servers might slap "Content-encoding: gzip" when they serve a file that happens to be compressed (without further compressing it): they might do this even if client did not ask for compression
* if response includes "Content-encoding: gzip", requests will decompress (even if explicitly did not request compression)
* the application can workaround this by reading(non-decompressed) response.raw but obviously you can only do that if you 100% know that you did not request compression -- otherwise it makes no sense

lukpueh replied:
Thanks for the insight. Would you mind adding a short note about that to the commit message, where you say "Don't use response.raw in RequestsFetcher"?

jku commented:
modified commit message with an explanation of why we no longer use requests.Response.raw (which is really a urllib3.response.HTTPResponse)
